src/request/sign-swap/SignSwap.js

```diff
@@ -0,0 +1,492 @@
+                return;
+            }
+            reject(new Errors.CoreError(e));
+            return;
+        }
+        if (!key) {
+            reject(new Errors.KeyNotFoundError());
+            return;
+        }
+
+        const btcKey = new BitcoinKey(key);
+
+        /** @type {{nim: string, btc: string[]}} */
+        const privateKeys = {};
+
+        if (request.fund.type === 'NIM') {
+            const privateKey = key.derivePrivateKey(request.fund.keyPath);
+            privateKeys.nim = privateKey.toHex();
+        }
+
+        if (request.fund.type === 'BTC') {
+            const keyPairs = request.fund.inputs.map(input => btcKey.deriveKeyPair(input.keyPath));
+            const privKeys = keyPairs.map(keyPair => Nimiq.BufferUtils.toHex(
+                /** @type {Buffer} */ (keyPair.privateKey),
+            ));
+            privateKeys.btc = privKeys;
+
+            if (request.fund.changeOutput) {
+                // Calculate, validate and store output address
+                const address = btcKey.deriveAddress(request.fund.changeOutput.keyPath);
+                if (request.fund.changeOutput.address && request.fund.changeOutput.address !== address) {
+                    throw new Errors.InvalidRequestError(
+                        'Given address is different from derived address for change output',
+                    );
+                }
+                request.fund.changeOutput.address = address;
+            }
+
+            // Calculate and store refund address from refundKeyPath
+            request.fund.refundAddress = btcKey.deriveAddress(request.fund.refundKeyPath);
+        }
+
+        if (request.redeem.type === 'NIM') {
+            const privateKey = key.derivePrivateKey(request.redeem.keyPath);
+            privateKeys.nim = privateKey.toHex();
+        }
+
+        if (request.redeem.type === 'BTC') {
+            const keyPairs = [btcKey.deriveKeyPair(request.redeem.input.keyPath)];
+            const privKeys = keyPairs.map(keyPair => Nimiq.BufferUtils.toHex(
+                /** @type {Buffer} */ (keyPair.privateKey),
+            ));
+            privateKeys.btc = privKeys;
+
+            // Calculate, validate and store output address
+            const address = btcKey.deriveAddress(request.redeem.output.keyPath);
+            if (request.redeem.output.address && request.redeem.output.address !== address) {
+                throw new Errors.InvalidRequestError(
+                    'Given address is different from derived address for output',
+                );
+            }
+            request.redeem.output.address = address;
+        }
+
+        try {
+            sessionStorage.setItem(
+                SignSwapApi.SESSION_STORAGE_KEY_PREFIX + request.swapId,
+                JSON.stringify({
+                    keys: privateKeys,
+
+                    // Serialize request to store in SessionStorage
+                    request: {
+                        ...request,
```

Comment:
In my opinion a little overboard with inlining.

i.e
```js
let copiedRequest = Object.assign({}, request);
if (copiedRequest.fund.type === 'NIM') {
    copiedRequest.fund.transaction = request.fund.transaction.toPlain();
}
```
Or something along those lines would have worked too. Potentially even in place. In my opinion is easier to read and shorter, as well.

Reply:
I was actually quite proud of this construct :laughing: